libnvme v2

igaw opened the pull request:
PR to trigger CI build

igaw committed:
map: merge all symbols into 2.0 section

Merge all existing symbols into a new 2.0 section.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: remove deprecated functions

Get rid of deprecated functions.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
src: introduce link handle

Abstract the file descriptor handle to a more abstract type so it's
possible to use different 'backends' to send/receive commands, e.g.
via MI instead direct.

The name 'link' was chosen because this name is not in use and thus we
avoid overloading existing names, such ctrl. The specification is using
'link' to describe the connection on a PCI bus which is kind of matching
how the handle is used in the library. It is the means to talk to a
controller.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
api-types: remove unused fd members

There is no user left for the file descriptor in the args structs,
thus remove it.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
linux: add nvme_link_get_fd

When the user provides their own implementation of the passtrhu
function, it is necessary to use the file descriptor for the ioctl call.
Thus allow to retrieve the file descriptor from the link handle.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
src: merge libnvme-mi into libnvme

libnvme-mi was added after the initial 1.x release and it was decided to
ship it independ of libnvme. With the upcoming 2.x work, merge the two
libraries so it's possible to use common code, that is unify the APIs.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
mi: replace dedicated MI API with generic one

With the introduction of the link handle, it's possible to use the
generic nvme API. The transport selection happens on lower levels
depending on the link object type.

Note, the tests rely on using the ioctl interface and not the io_uring.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
linux: add nvme_link_get_name

nvme-cli wants to retrieve the name of the device, add a getter for
this.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
linux: add nvme_link_is_{blk|char}dev

The nvme_link_t type is used for block device and char device. User
application sometimes need to figure out which type of device is used.
Because libnvme already knows the type of device expose this. This
avoids tracking of the device type in the user application.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: add alignment and length check to nvme_fw_download

The MI code has an alignment and length check which the direct APIs are
missing. Without these checks the download tests for MI will fail after
using the generic APIs.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
mi: replace nvme_mi_ctrl_t with nvme_link_t

nvme_link_t in place, it's possible to replace nvme_mi_ctrl_t with the
generic transport abstraction type.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
api-types: move nvme_root_t to new place

The root object will not only be used by the tree API, instead it will
be used as default context for the complete API. This addresses the
shortcoming that 'global' variables such as the logging infrastructure
should not be library global variables.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
linux: add MI support to nvme_{open|close}

Extend nvme_open and nvme_close to handle MI devices.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: add support for MI passthru

Hook up the MI backend to the generic commands API.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
linux: add nvme_link_is_{direct|mi}

The nvme_link_t type is used to abstract the transport type. User
application sometimes need to figure out which type of device is used.
Because libnvme already knows the type of device expose this. This
avoids tracking of the device type in the user application.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
src: return error codes directly

The v1 design used the POSIX way to return errors, return -1 with errno
set.

Replace this design by returning directly the error code. That means non
trivial function have to return any data via a pointer so the function
is able to return proper error codes.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_get_log use nvme_passthru_cmd directly

Refactor nvme_get_log and nvme_get_log_page to use struct
nvme_passthru_cmd directly and drop the struct nvme_get_log_args.

The struct args have are ABI thus it's hard to change it after a
release. The nvme_passthru_cmd buffer is a generic buffer and it's
possible to use the setters directly on the buffer in an inline
function. Besides a reduced API to maintain it also generates better
code and avoids one call less when issue a command.

[ikegami.t: update nvme_uring_cmd_amdin_passthru_async]
Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: reorder nvme_get_log and nvme_get_log_page

The next change will add a static inline implementation of nvme_get_log
which uses nvme_get_log_page. Thus move the nvme_get_log_page function
up so it can be used directly.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

MaisenbacherD committed:
ioctl: nvme_set_features use nvme_passthru_cmd directly

Drop struct nvme_set_features_args.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>
[wagi: remove libnvme.map entries
       reorder fid and sv argument]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: add nvme_identify_partial

MI doesn't access the complete identify data struct and thus
only transfers a port of it. Add a generic version of this
function.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
mi: remove nvme_mi_admin_identify{_partial}

Use the generic nvme_identify{_partial} interface and drop
the MI specific version.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_identify use nvme_passthru_cmd directly

[dennis: fix example]
Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>
Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_format_nvm use nvme_passthru_cmd directly

Drop struct nvme_format_nvm_args.

[dennis: merge lbafl and lbafu into one argument]
Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>
Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_ns_mgmt use nvme_passthru_cmd directly

Drop struct nvme_ns_mgmt_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_ns_attach use nvme_passthru_cmd directly

Drop struct nvme_ns_attach_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_ns_download use nvme_passthru_cmd directly

Drop struct nvme_fw_download_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_fw_commit use nvme_passthru_cmd directly

Drop struct nvme_fw_commit_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_security_send use nvme_passthru_cmd directly

Drop struct nvme_security_send_args.

[dennis: merge spsp0 and spsp1 into one argument]
Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_dsm use nvme_passthru_cmd directly

Drop struct nvme_dsm_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reorder arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_io use nvme_passthru_cmd directly

Drop struct nvme_io_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: add dsm and io commands dword fields

To use the fields in nvme_io() and nvme_dsm() functions.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: update API due to nvme_io API rework]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_copy use nvme_passthru_cmd directly

Drop struct nvme_copy_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reorder and adding missing arguments
       update storage/app tag interface]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
test: fix rae and lsp handling in log page tests

The compiler warns with

../test/ioctl/logs.c: In function ‘test_get_log_reachability_groups’:
../test/ioctl/logs.c:725:28: warning: enum constant in boolean context [-Wint-in-bool-context]
  725 |                          (!!TEST_LSP << 8) | (!!TEST_RAE << 15) |
      |                            ^
../test/ioctl/logs.c:733:70: warning: enum constant in boolean context [-Wint-in-bool-context]
  733 |         err = nvme_get_log_reachability_groups(test_link, TEST_RAE, !!TEST_LSP,
      |                                                                      ^

Update the tests so the types are correctly handled. It also addresses
the problem with the wrong lsp value testing.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_resv_acquire use nvme_passthru_cmd directly

Drop struct nvme_resv_acquire_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reorder and missing arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_resv_register use nvme_passthru_cmd directly

Drop struct nvme_resv_register_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reorder and added missing arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_resv_release use nvme_passthru_cmd directly

Drop struct nvme_resv_release_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_resv_report use nvme_passthru_cmd directly

Drop struct nvme_resv_report_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reorder and added missing arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_io_mgmt_recv use nvme_passthru_cmd directly

Drop struct nvme_io_mgmt_recv_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reorder arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_io_mgmt_send use nvme_passthru_cmd directly

Drop struct nvme_io_mgmt_send_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reorder arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_zns_mgmt_send use nvme_passthru_cmd directly

Drop struct nvme_zns_mgmt_send_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reorderd and added missing arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_zns_mgmt_recv use nvme_passthru_cmd directly

Drop struct nvme_zns_mgmt_recv_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: rename and reorder arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_zns_append use nvme_passthru_cmd directly

Drop struct nvme_zns_append_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[dennis: add end-to-end PI tags]
Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>
[wagi: reorder and added missing arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_dim_send use nvme_passthru_cmd directly

Drop struct nvme_dim_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
fabrics: fix nvme_get_adrfam function description parameter ctrl as c

Since the warning message output by kernel-doc-check.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

MaisenbacherD committed:
ioctl: nvme_get_features use nvme_passthru_cmd directly

Drop struct nvme_get_features_args.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>
[wagi: reordered/renamed arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_lm_cdq use nvme_passthru_cmd directly

Drop struct nvme_lm_cdq_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reorder arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_lm_track_send use nvme_passthru_cmd directly

Drop struct nvme_lm_track_send_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reordered arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_lm_migration_send use nvme_passthru_cmd directly

Drop struct nvme_lm_migration_send_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reordered arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_lm_migration_recv use nvme_passthru_cmd directly

Drop struct nvme_lm_migration_recv_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
test/zns: fix kernel-doc-check warnings

Since the following warnings output.

$ ./scripts/kernel-doc-check test/zns.c
test/zns.c:3: warning: This comment starts with '/**', but isn't a kernel-doc comment. Refer Documentation/doc-guide/kernel-doc.rst
test/zns.c:10: warning: This comment starts with '/**', but isn't a kernel-doc comment. Refer Documentation/doc-guide/kernel-doc.rst

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_security_receive use nvme_passthru_cmd directly

Drop struct nvme_security_receive_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reordered arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
examples: fix mi-mctp.c kernel-doc-check warning

examples/mi-mctp.c:3: warning: This comment starts with '/**', but isn't a kernel-doc comment. Refer Documentation/doc-guide/kernel-doc.rst

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_get_lba_status use nvme_passthru_cmd directly

Drop struct nvme_get_lba_status_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: reordering arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_directive_send use nvme_passthru_cmd directly

Drop struct nvme_directive_send_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: updated nvme_directive_send_id_endir
       reordered arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_directive_recv use nvme_passthru_cmd directly

Drop struct nvme_directive_recv_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: removed libnvme.map entry
       reordered arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_virtual_mgmt use nvme_passthru_cmd directly

Drop struct nvme_virtual_mgmt_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: remove libnvme.map entry]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_dev_self_test use nvme_passthru_cmd directly

Drop struct nvme_dev_self_test_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: removed libnvme.map entry]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_sanitize_nvm use nvme_passthru_cmd directly

Drop struct nvme_sanitize_nvm_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: remove libnvme.map entry
       reorder and rename arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
test: fix mi.c kernel-doc-check warning

To fix the warning message below.
  test/mi.c:3: warning: This comment starts with '/**', but isn't a kernel-doc comment. Refer Documentation/doc-guide/kernel-doc.rst

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_get_property use nvme_passthru_cmd directly

Drop struct nvme_get_property_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: remove libnvme.map entry]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_set_property use nvme_passthru_cmd directly

Drop struct nvme_set_property_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: remove libnvme.map entry]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_lockdown use nvme_passthru_cmd directly

Drop struct nvme_lockdown_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: remove libnvme.map entry]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
examples: do not use "/**" opening comment mark

Since checked by kernel-doc-check then the warning message output.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: fix get_log_partial to use uring for link type direct

Since uring not supported for link type mi.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: refactor nvme_get_log_partial uring

This is to maintain nvme_get_log_partial easily.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_capacity_mgmt use nvme_passthru_cmd directly

Drop struct nvme_capacity_mgmt_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>
[wagi: remove libnvme.map entry
       reordered/renamed arguments]
Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: remove unused enum features

There is no user left of this type in ioctl.c. Remove it.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

MaisenbacherD committed:
ioctl: introduce SHIFT MASK enum for nvme_capacity_mgmt

Replace hardcoded MASK and SHIFT values with enum entries for
nvme_capacity_mgmt.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

MaisenbacherD committed:
ioctl: introduce SHIFT MASK enum for nvme_lockdown

Replace hardcoded MASK and SHIFT values with enum entries for
nvme_lockdown.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

MaisenbacherD committed:
ioctl: introduce SHIFT MASK enum for nvme_io_mgmt_recv

Replace hardcoded MASK and SHIFT values with enum entries for
nvme_io_mgmt_recv.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

MaisenbacherD committed:
ioctl: introduce SHIFT MASK enum for nvme_resv_*

Replace hardcoded MASK and SHIFT values with enum entries for
nvme_resv_*.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

igaw committed:
ioctl: update nvme_io related APIs

This will be merged into the previous patch after reviewing.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

MaisenbacherD committed:
ioctl: add return value for nvme_set_features_iocs_profile

The spec specifies that this command returns the currently selected
IOCSCI upon successful completion.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

MaisenbacherD committed:
ioctl: Rename NVME_NVM_CDW* SHIFT and MASK enum members

Renaming NVME_NVM_CDW* SHIFT and MASK enum members to
NVME_IOCS_COMMON_CDW*.
The ZNS IO command set (IOCS) has the same fields and offsets for the CDWs
in question.
Replacing STCR with two enum entries for CDW12 bit 25 as this bit can only
be Reserved, DEAC or PIREMAP. STCR is part of the copy command which has
its own enum members for CDW12.

The NVME_IOCS_COMMON_CDW* entries cover now
nvme_read
nvme_write
nvme_compare
nvme_write_zeros
nvme_write_uncorrectable
nvme_zns_append

This is in preparation for the following commit which introduces the
shift masks for the zone append command.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

MaisenbacherD committed:
ioctl: introduce SHIFT MASK enum for nvme_zns_append

Replace hardcoded MASK and SHIFT values with enum entries for
nvme_zns_append.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

igaw commented:
Just fixed many checkpath warnings/errors, added commit messages, fixed style issues and a real errors. This means the series is fairly clean at least in the context of checkpath

MaisenbacherD commented:
Closing this, as this was just a means to get the CI to build.

MaisenbacherD commented:
> Just fixed many checkpath warnings/errors, added commit messages, fixed style issues and a real errors. This means the series is fairly clean at least in the context of checkpath

Darn it! I didn't read your latest commend before closing this. Sorry I am reopening this.

MaisenbacherD commented:
@igaw Ohh! Very smart to keep this PR open, that way we always trigger the build action with changes on the libnvme2 branch.

igaw committed:
types: update include order for types

The api-types.h include file provided struct args definition. These are
all gone. But some basic types have been added in the meantime, which
need to be present when using the types.h header.

Signed-off-by: Daniel Wagner <wagi@kernel.org>
